Assets/Scripts/Entity/Enemy/Heart.cs

```diff
@@ -21,31 +22,41 @@ private void OnCollisionEnter(Collision other)
                 {
                     player.ChangeHealthBy(player.MaxHealth - player.Health);
                     player.ChangeGoldenHealthBy(Health);
-                    StartCoroutine(DestroyWithDelay());
+                    DestroyHeartButPlayAudio();
                 }
                 else if (player.GoldenHealth != player.MaxGoldenHealth)
                 {
                     player.ChangeHealthBy(player.MaxHealth - player.Health);
                     player.ChangeGoldenHealthBy(player.MaxGoldenHealth - player.Health);
-                    StartCoroutine(DestroyWithDelay());
+                    DestroyHeartButPlayAudio();
                 }
             }
             else if(player.Health != player.MaxHealth)
             {
                 if ((player.Health + Health) <= player.MaxHealth)
                 {
                     player.ChangeHealthBy(Health);
-                    StartCoroutine(DestroyWithDelay());
+                    DestroyHeartButPlayAudio();
                 }
                 else if (player.Health != player.MaxHealth)
                 {
                     player.ChangeHealthBy(player.MaxHealth - player.Health);
-                    StartCoroutine(DestroyWithDelay());
+                    DestroyHeartButPlayAudio();
                 }
             }
         }
     }
 
+    private void DestroyHeartButPlayAudio()
+    {
+        /**
+            Due to some logical/design oversight, we have to hide the heart
+            and disable the collider immediately, but not destroy it
+            because we want the audio to play (Which is a child of this object).
```

Comment:
I don't see how this code fixes the race condition, or what the race condition actually was. The game object never got destroyed until after the audio played.

Reply:
The race condition was between the OnCollisionEnter method being called, and the hearts collider being disabled. Due to the disabling of the collider (`HideHeart` method) being within the Coroutine, there was no guarantee that the heart would be disabled/hid in the current frame or even the next frame. Allowing for the possibility of the `OnCollisionEnter` method to be called again.

The issue stems from the fact that the heart is not destroyed immediately on collision, since it needs to fully play the audio first.